### DISCUSSION TITLE
Github actions: run multiple jobs in the same docker

### DISCUSSION THREAD

**Discussion by w3ntao** [community/community#26545] [Actions]
<p>I’m learning to deploy github actions to run multiple jobs with docker, and this is what I have so far:</p>
<p>github actions yml file is shown as follow. There are 2 jobs: <code>job0</code> builds docker with <code>Dockerfile0</code> and <code>job1</code> builds docker with <code>Dockerfile1</code>.</p>
<pre><code class="lang-auto"># .github/workflows/main.yml
name: docker CI

on: push

jobs:
  job0:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Build and Run
      run: docker build . --file Dockerfile0 --tag job0

  job1:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Build and Run
      run: docker build . --file Dockerfile1 --tag job1
</code></pre>
<p><code>Dockerfile0</code> and <code>Dockerfile1</code> share basically the same content, except for the argument in the last line:</p>
<pre><code class="lang-auto">FROM ubuntu:20.04

ADD . /docker_ci
RUN apt-get update -y
RUN apt-get install -y ... ...

WORKDIR /docker_ci
RUN python3 script.py &lt;arg&gt;
</code></pre>
<p>I wonder, can I build a docker for the 1st job, and then invoke multiple jobs execute command within the docker built from the 1st job? Thus I don’t have to keep multiple Dockerfile and save some docker building time.</p>
<p>It would be better to build my docker locally from Dockerfile so I hope to avoid using <code>container</code> from docker hub.</p>
<p><a href="https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-docker-actions" rel="noopener nofollow ugc">runs-for-docker-actions</a> looks relevant but I have trouble finding example deploying action locally (without publishing).</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="wentaocn" data-post="1" data-topic="179664">
<div class="title">
<div class="quote-controls"></div>
 wentaocn:</div>
<blockquote>
<p>I wonder, can I build a docker for the 1st job, and then invoke multiple jobs execute command within the docker built from the 1st job? Thus I don’t have to keep multiple Dockerfile and save some docker building time.</p>
</blockquote>
</aside>
<p>I’m not exactly sure what you’re trying to do, do you want to use a container built in one job as the build environment (job container, even?) for later jobs?</p>
<p>One thing I can definitely say is that jobs in GitHub hosted runners each get a fresh VM, so if you build a container in one job and want to use it in another you have to push it to a registry, or store the image somewhere else.</p>

**Comment by w3ntao** [comment]
<blockquote>
<p>jobs in GitHub hosted runners each get a fresh VM</p>
</blockquote>
<p>This is certainly something I didn’t know. Thanks, problem solved.</p>
